### DISCUSSION TITLE
Multibinding a style class in code

### DISCUSSION THREAD

**Discussion by DarkMortainius** [AvaloniaUI/Avalonia#15711] [Q&A] [answered]
I'm trying to create a data template in code for a TreeDataGrid, and I want to have my data cells have a style class added if the user has write access to a specific data cell. Essentially, I want to turn this XAML into code, and I'm not sure how (specifically the part concerning the style class). I've already got the converter working for other items. I just need to convert this portion into code

```XAML
<DataTemplate x:DataType="datapoint:DataPoint">                        
    <Label Content="Sample Text">
        <Label.Classes.Edit>
            <MultiBinding Converter="{StaticResource WriteAccessConverter}">
                <Binding Path=""/>
                <Binding Path="$parent[TreeDataGrid].((vm:ParameterTabViewModel)DataContext).ActiveUser"/>
            </MultiBinding>
        </Label.Classes.Edit>
    </Label>
</DataTemplate>
```

This is the closest I've been able to get (adding the class without the conditional binding)
```c#
var cellTemplate = new FuncDataTemplate<DataRow>((_, _) => {
    var label = new Label { [!Label.ContentProperty] = new Binding(bindingString + ".Text") };
    label.Classes.Add("Edit");
    return new ContentControl() { Content = label };                    
});
```

**Comment by stevemonaco** [comment]
Remove the `Label.` part. eg.

```xml
  <Window.Styles>
    <Style Selector="TextBlock.big">
      <Setter Property="FontSize" Value="36" />
    </Style>
  </Window.Styles>

  <TextBlock x:Name="text" Text="Testing MultiBinding">
    <Classes.big>
      <MultiBinding Converter="{x:Static BoolConverters.And}">
        <Binding Path="IsA" />
        <Binding Path="IsB" />
      </MultiBinding>
    </Classes.big>
  </TextBlock>
```

You also can't use syntax the same way you do in the `{Binding}` `MarkupExtension`. You need to fill out the individual properties when using `<Binding />` directly, AFAIK. ie. You need to use `RelativeSource` instead of `$parent[TreeDataGrid]` and whatnot. If you were trying to do a self-binding in the first `<Binding />`, then that might still be unsupported. You might check out issues regarding that.

You could also write your own `MarkupExtension` instead which can support bindings (via `{Binding}`) as parameters.

**Reply by DarkMortainius** [reply]
Ok, so my syntax was a little off. But that wasn't what my question was. I'm trying to do this in C# code, not XAML, and I'm not sure how. I can't find any examples of assigning bindings to classes in C# code.

**Reply by stevemonaco** [reply, marked as answer]
I see, sorry about that. You need to work with `StyledElement.BindClass` which is an extension method in the `Avalonia` namespace. Converting my example...

```cs
public MainWindow()
{
    InitializeComponent();

    var mb = new MultiBinding()
    {
        Bindings = [new Binding("IsA"), new Binding("IsB")],
        Converter = BoolConverters.And
    };

    text.BindClass("big", mb, null);
}
```

```xml
<Window.Styles>
  <Style Selector="TextBlock.big">
    <Setter Property="FontSize" Value="36" />
  </Style>
</Window.Styles>

<Grid RowDefinitions="auto,*">
  <TextBlock x:Name="text" Text="Testing MultiBinding" />
</Grid>
```

**Reply by DarkMortainius** [reply]
That looks exactly like what I need, I'll give it a try. Thank you!

**Reply by DarkMortainius** [reply]
Do you know what anchor is supposed to be? It's non-nullable, so the compiler is complaining if I try to pass null, and the documentation is no help.

[BindClass(IStyledElement, string, IBinding, object)](https://reference.avaloniaui.net/api/Avalonia/StyledElementExtensions/D321134D)

**Reply by DarkMortainius** [reply]
After digging into the code, it looks like it doesn't matter what anchor is, it never actually gets used by anything...
Not sure why it even needs to be there.

**Reply by stevemonaco** [reply]
That's what I got out of it, too. I think anchor was used in the old `IBinding` system in `SourceNode` to do something with the binding `Source`, but now it seems to have faded away with `IBinding2` and can't be removed until v12 because it's a breaking change.

I'll post an issue about changing it to `object?` instead. Maybe it can get into 11.1.

Also, the reference docs site shouldn't be used. It's still on v0.10.x docs as there are some technical issues with re-generating them. You'll need to check source.

**Comment by hydrolizer** [comment]
My solution:
```
using Avalonia;
using Avalonia.Xaml.Interactivity;

namespace Attached;

public sealed class StyleClassBinding: Behavior<StyledElement>
{
  public static readonly StyledProperty<bool> IsSuitableProperty
    = AvaloniaProperty.Register<StyleClassBinding, bool>(nameof(IsSuitable));
  
  public bool IsSuitable
  {
    get => GetValue(IsSuitableProperty);
    set => SetValue(IsSuitableProperty, value);
  }

  public string? ClassName { get; set; }

  protected override void OnAttached()
  {
    base.OnAttached();
    if (string.IsNullOrWhiteSpace(ClassName)) return;
    AssociatedObject?.BindClass(ClassName, this.GetObservable(IsSuitableProperty).ToBinding(), null!);
  }
}
```
Example of use:
```
<Grid RowDefinitions="*,Auto">
  <StyledElement.Resources>
    <local:BooleanAndConverter x:Key="BooleanAndConverter" />
  </StyledElement.Resources>
  <StyledElement.Styles>
    <Style Selector="Border#TestBorder">
      <Setter Property="Background" Value="Aqua" />
    </Style>
    <Style Selector="Border#TestBorder.allChecked">
      <Setter Property="Background" Value="Green" />
    </Style>
  </StyledElement.Styles>
  <Border x:Name="TestBorder">
    <i:Interaction.Behaviors>
      <local:StyleClassBinding ClassName="allChecked">
        <local:StyleClassBinding.IsSuitable>
          <MultiBinding Converter="{StaticResource BooleanAndConverter}">
            <Binding ElementName="FirstCheckBox" Path="IsChecked" />
            <Binding ElementName="SecondCheckBox" Path="IsChecked" />
          </MultiBinding>
        </local:StyleClassBinding.IsSuitable>
      </local:StyleClassBinding>
    </i:Interaction.Behaviors>
  </Border>
  <StackPanel Grid.Row="1" Margin="0,5,0,0">
    <CheckBox x:Name="FirstCheckBox" Content="First" />
    <CheckBox x:Name="SecondCheckBox" Content="Second" />
  </StackPanel>
</Grid>
```
